### PR TITLE
Optimize protobuf encoder and iterator for various workloads

### PR DESCRIPTION
**Pull request by richardartoul**
**What this PR does / why we need it**:
Remove `jhump/protoreflect` from the hot path as much as possible, plus some other optimizations and cleanup.

Results:

1. 10x improvement in encoding and decoding for non-custom fields.
2. 2x improvement in encoding for custom fields.
3. 4x improvement in decoding for custom fields (original goal).

before
```
goos: darwin
goarch: amd64
pkg: github.com/m3db/m3/src/dbnode/encoding/proto
BenchmarkEncoder/with_non_custom_encoded_fields_enabled-8         	    1000	   1432754 ns/op	  374435 B/op	    8177 allocs/op
BenchmarkEncoder/with_non_custom_encoded_fields_disabled-8        	   10000	    111547 ns/op	   13430 B/op	       0 allocs/op
BenchmarkIterator/with_non_custom_encoded_fields_enabled-8        	    1000	   1274402 ns/op	  323059 B/op	    8196 allocs/op
BenchmarkIterator/with_non_custom_encoded_fields_disabled-8       	   10000	    203996 ns/op	   21603 B/op	     910 allocs/op
```

after
```
goos: darwin
goarch: amd64
pkg: github.com/m3db/m3/src/dbnode/encoding/proto
BenchmarkEncoder/with_non_custom_encoded_fields_enabled-8         	   10000	    119435 ns/op	   66221 B/op	       0 allocs/op
BenchmarkEncoder/with_non_custom_encoded_fields_disabled-8        	   20000	     67853 ns/op	   21176 B/op	       0 allocs/op
BenchmarkIterator/with_non_custom_encoded_fields_enabled-8        	   20000	     90630 ns/op	     671 B/op	      13 allocs/op
BenchmarkIterator/with_non_custom_encoded_fields_disabled-8       	   30000	     46309 ns/op	     595 B/op	      12 allocs/op
```

### CODE REVIEW

File: src/dbnode/encoding/m3tsz/float_encoder_iterator.go [hunk 1]
```diff
@@ -140,16 +140,13 @@ func (eit *FloatEncoderAndIterator) readNextFloat(stream encoding.IStream) error
...
 		return err
 	}
 
-	numMeaningfulBits, err := stream.ReadBits(6)
-	if err != nil {
-		return err
-	}
-	numMeaningfulBits = numMeaningfulBits + 1
+	numLeadingZeros := (numLeadingZeroesAndNumMeaningfulBits & 4032) >> 6
+	numMeaningfulBits := (numLeadingZeroesAndNumMeaningfulBits & 63) + 1
```
Comment (robskillington) [inline src/dbnode/encoding/m3tsz/float_encoder_iterator.go]: Can we make these two numbers consts? 

i.e. 
```go
const (
  bits12To6Set = 4032 // 1111 1100 0000
  bits6To0Set = 63 // 0011 1111
)
```

File: src/dbnode/encoding/proto/common.go [hunk 2]
```diff
@@ -141,11 +143,17 @@ var (
 	}
 )
 
+type marshaledField struct {
```
Comment (robskillington) [inline src/dbnode/encoding/proto/common.go]: nit: `marshaledField` should be `marshalledField` I believe

File: src/dbnode/encoding/proto/common.go [hunk 3]
```diff
@@ -141,11 +143,17 @@ var (
...
+	fieldNum  int32
+	marshaled []byte
```
Comment (robskillington) [inline src/dbnode/encoding/proto/common.go]: nit: Should be `marshalled` I believe

File: src/dbnode/encoding/proto/custom_marshal.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	"math"
+
+	"github.com/golang/protobuf/proto"
+)
+
+type customFieldMarshaler interface {
+	encFloat64(tag int32, x float64)
+	encFloat32(tag int32, x float32)
+	encInt32(tag int32, x int32)
+	encSInt32(tag int32, x int32)
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_marshal.go]: I assume the S variants are `varint` or `short`?  Maybe use `encVarInt32` or `encShortInt32` instead?  

The single letter S is strange and not as clear the delineation while reading the API.

File: src/dbnode/encoding/proto/custom_marshal_test.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+	marshaler := newCustomMarshaler()
+	for _, tc := range testCases {
+		marshaler.reset()
+		customMarshalVL(t, marshaler, tc.message)
+
+		unmarshalM := dynamic.NewMessage(testVLSchema)
+		require.NoError(t, unmarshalM.Unmarshal(marshaler.bytes()))
+
+		require.True(t, dynamic.Equal(tc.message, unmarshalM))
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_marshal_test.go]: So you can run individual tests, etc it's best practice to wrap this in a `t.Run(...)`:
```go
for i, tc := range testCases {
  t.Run(fmt.Sprintf("my_test_%d", i), func (t *testing.T) {
    // test case innards
  })
}
```

Also helps if you need defer, etc in the body.

File: src/dbnode/encoding/proto/custom_marshal.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+package proto
+
+import (
+	"math"
+
+	"github.com/golang/protobuf/proto"
+)
+
+type customFieldMarshaler interface {
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_marshal.go]: `customFieldMarshaler` should be `customFieldMarshaller` I believe.

File: src/dbnode/encoding/proto/custom_marshal.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+	encUInt64(tag int32, x uint64)
+	encBool(tag int32, x bool)
+	encBytes(tag int32, x []byte)
+	encPartialProto(tag int32, x []byte)
+	bytes() []byte
+	setBytes(b []byte)
+	reset()
+}
+
+type customMarshaler struct {
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_marshal.go]: `customMarshaller`

File: src/dbnode/encoding/proto/custom_marshal.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+	buf *buffer
+}
+
+func newCustomMarshaler() customFieldMarshaler {
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_marshal.go]: `newCustomMarshaller`

File: src/dbnode/encoding/proto/custom_unmarshal.go [hunk 9]
```diff
@@ -68,18 +68,15 @@ func (u *customUnmarshaler) numNonCustomValues() int {
...
+func (u *customUnmarshaler) nonCustomFieldValues() []marshaledField {
 	return u.nonCustomValues
 }
 
 func (u *customUnmarshaler) unmarshal() error {
 	u.customValues = u.customValues[:0]
 
 	if u.nonCustomValues != nil {
-		u.nonCustomValues.Reset()
+		u.nonCustomValues = u.nonCustomValues[:0]
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_unmarshal.go]: Do you need to zero out each entry to avoid holding onto refs or does `marshalledField` here not contain references to longly lived pointer types?

Comment (robskillington) [reply src/dbnode/encoding/proto/custom_unmarshal.go]: Hm I see a `[]byte` in that struct that could be longly held onto.

File: src/dbnode/encoding/proto/custom_unmarshal.go [hunk 10]
```diff
@@ -96,29 +93,40 @@ func (u *customUnmarshaler) unmarshal() error {
...
 
 			var (
-				startIdx       = tagAndWireTypeStartOffset
-				endIdx         = u.decodeBuf.index
-				marshaledField = u.decodeBuf.buf[startIdx:endIdx]
+				startIdx  = tagAndWireTypeStartOffset
+				endIdx    = u.decodeBuf.index
+				marshaled = u.decodeBuf.buf[startIdx:endIdx]
 			)
 			// A marshaled Protobuf message consists of a stream of <fieldNumber, wireType, value>
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_unmarshal.go]: nit: `A marshalled ...`

File: src/dbnode/encoding/proto/custom_unmarshal.go [hunk 11]
```diff
@@ -96,29 +93,40 @@ func (u *customUnmarshaler) unmarshal() error {
...
 				return err
 			}
 
 			var (
-				startIdx       = tagAndWireTypeStartOffset
-				endIdx         = u.decodeBuf.index
-				marshaledField = u.decodeBuf.buf[startIdx:endIdx]
+				startIdx  = tagAndWireTypeStartOffset
+				endIdx    = u.decodeBuf.index
+				marshaled = u.decodeBuf.buf[startIdx:endIdx]
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_unmarshal.go]: nit: `marshalled = ...`

File: src/dbnode/encoding/proto/custom_unmarshal.go [hunk 12]
```diff
@@ -96,29 +93,40 @@ func (u *customUnmarshaler) unmarshal() error {
...
-			// marshaled message and as a result we can build up the nonCustomValues *dynamic.Message
-			// one field at a time by calling UnmarshalMerge() on sub-slices that contain a complete
-			// tuple.
-			if err := u.nonCustomValues.UnmarshalMerge(marshaledField); err != nil {
-				return err
+			// marshaled message and as a result we can build up the []marshaledField one field at
+			// a time.
+			found := false
+			if fd.IsRepeated() {
+				// TODO: Leave comment
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_unmarshal.go]: Still need to add comment?

File: src/dbnode/encoding/proto/custom_unmarshal.go [hunk 13]
```diff
@@ -96,29 +93,40 @@ func (u *customUnmarshaler) unmarshal() error {
...
+				for i := range u.nonCustomValues {
+					val := u.nonCustomValues[i]
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_unmarshal.go]: nit: Perhaps `for i, val := range u.nonCustomValues` to remove this line?

File: src/dbnode/encoding/proto/custom_unmarshal.go [hunk 14]
```diff
@@ -332,14 +344,15 @@ func unmarshalSimpleField(fd *desc.FieldDescriptor, v uint64) (unmarshalValue, e
...
-	}
-
 	u.schema = schema
 	u.numNonCustom = 0
 	u.customValues = u.customValues[:0]
+
+	for i := range u.nonCustomValues {
+		u.nonCustomValues[i] = marshaledField{}
+	}
+	u.nonCustomValues = u.nonCustomValues[:0]
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_unmarshal.go]: You could make this an member method so in the other place you zero out you remember to zero the actual fields?  i.e. `u.resetNonCustomValues()`

File: src/dbnode/encoding/proto/custom_unmarshal.go [hunk 15]
```diff
@@ -332,14 +344,15 @@ func unmarshalSimpleField(fd *desc.FieldDescriptor, v uint64) (unmarshalValue, e
...
 
 func (u *customUnmarshaler) resetAndUnmarshal(schema *desc.MessageDescriptor, buf []byte) error {
-	if schema != u.schema {
-		u.nonCustomValues = dynamic.NewMessage(schema)
-		u.nonCustomValues.Reset()
-	}
-
 	u.schema = schema
 	u.numNonCustom = 0
 	u.customValues = u.customValues[:0]
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_unmarshal.go]: Does `customValues` also need its entries reset? (does it contain any pointer types?)

File: src/dbnode/encoding/proto/custom_unmarshal_test.go [hunk 16]
```diff
@@ -151,11 +151,11 @@ func TestCustomFieldUnmarshaler(t *testing.T) {
...
-			m := unmarshaler.nonCustomFieldValues()
-			assertAttributesEqual(
-				t,
-				tc.attributes,
-				m.GetFieldByName("attributes").(map[interface{}]interface{}))
+			// m := unmarshaler.nonCustomFieldValues()
+			// assertAttributesEqual(
+			// 	t,
+			// 	tc.attributes,
+			// 	m.GetFieldByName("attributes").(map[interface{}]interface{}))
```
Comment (robskillington) [inline src/dbnode/encoding/proto/custom_unmarshal_test.go]: Should this still be commented out?

File: src/dbnode/encoding/proto/encoder.go [hunk 17]
```diff
@@ -657,7 +652,8 @@ func (enc *Encoder) encodeBoolValue(i int, val bool) {
 	}
 }
 
-func (enc *Encoder) encodeProtoValues(m *dynamic.Message) error {
+// TODO: Rename encodeNonCustomValues
```
Comment (robskillington) [inline src/dbnode/encoding/proto/encoder.go]: Still relevant?

File: src/dbnode/encoding/proto/encoder.go [hunk 18]
```diff
@@ -670,77 +666,71 @@ func (enc *Encoder) encodeProtoValues(m *dynamic.Message) error {
...
 		// Only want to skip encoding if nothing has changed AND we've already
 		// encoded the first message.
 		enc.stream.WriteBit(opCodeNoChange)
 		return nil
 	}
 
-	marshaled, err := m.MarshalAppend(enc.marshalBuf[:0])
-	if err != nil {
-		return fmt.Errorf("%s error trying to marshal protobuf: %v", encErrPrefix, err)
+	marshalBuf := enc.marshalBuf[:0]
```
Comment (robskillington) [inline src/dbnode/encoding/proto/encoder.go]: Hm, seems you reset the `enc.marshalBuf` back to the value of this local var.  Perhaps remove the `marshalBuf` var altogether and just use `enc.marshalBuf` below?

File: src/dbnode/encoding/proto/encoder_test.go [hunk 19]
```diff
@@ -24,6 +24,7 @@ import (
 	"testing"
 	"time"
 
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
```
Comment (robskillington) [inline src/dbnode/encoding/proto/encoder_test.go]: Hm this import doesn't belong here yeah?

Linter is definitely getting way more flakey, noticed a lot of these.

File: src/dbnode/encoding/proto/iterator.go [hunk 20]
```diff
@@ -26,6 +26,7 @@ import (
 	"io"
 	"math"
 
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
```
Comment (robskillington) [inline src/dbnode/encoding/proto/iterator.go]: Same here.

File: src/dbnode/encoding/proto/iterator.go [hunk 21]
```diff
@@ -51,18 +51,19 @@ var (
...
-	lastIterated           *dynamic.Message
-	lastIteratedProtoBytes []byte
-	byteFieldDictLRUSize   int
+	nsID                 ident.ID
+	opts                 encoding.Options
+	err                  error
+	schema               *desc.MessageDescriptor
+	schemaDesc           namespace.SchemaDescr
+	stream               encoding.IStream
+	marshaler            customFieldMarshaler
```
Comment (robskillington) [inline src/dbnode/encoding/proto/iterator.go]: nit: Should be `marshaller` I believe

File: src/dbnode/encoding/proto/iterator.go [hunk 22]
```diff
@@ -418,17 +435,20 @@ func (it *iterator) readProtoValues() error {
...
 
+	// TODO: Reset
 	m := dynamic.NewMessage(it.schema)
 	err = m.Unmarshal(unmarshalBytes)
 	if err != nil {
 		return fmt.Errorf("error unmarshaling protobuf: %v", err)
 	}
 
-	for _, field := range m.GetKnownFields() {
+	for _, marshaledField := range it.protoFields {
```
Comment (robskillington) [inline src/dbnode/encoding/proto/iterator.go]: nit: Should be `marshalledField` I believe.  Maybe just call it `field`?

File: src/dbnode/encoding/m3tsz/float_encoder_iterator.go (hunk 1 above)
Comment (richardartoul) [reply src/dbnode/encoding/m3tsz/float_encoder_iterator.go]: sure, but replaced "set" with "mask"

File: src/dbnode/encoding/proto/custom_marshal.go (hunk 4 above)
Comment (richardartoul) [reply src/dbnode/encoding/proto/custom_marshal.go]: Yeah I can see how that would be confusing. It actually stands for "signed" and these names are basically verbatim what they're called in in the Protobuf specification.

I'll add a comment

File: src/dbnode/encoding/proto/iterator.go [hunk 23]
```diff
@@ -185,6 +185,14 @@ func (it *iterator) Next() bool {
 				it.err = fmt.Errorf("%s error reading custom fields schema: %v", itErrPrefix, err)
 				return false
 			}
+
+			// When the encoder changes its schema it will reset all of its nonCustomFields state
+			// which means that the iterator needs to do the same to keep them synchronized at
```
Comment (Haijuncao) [inline src/dbnode/encoding/proto/iterator.go]: @richardartoul  you mentioned there is some bug around schema change, is this it? 
I saw you made some changes for encoder.SetSchema but later reverted it.

Which test caught the bug?

Comment (richardartoul) [reply src/dbnode/encoding/proto/iterator.go]: @Haijuncao `TestBufferSnapshot` caught it. I added a call to `SetSchema` in `RoundTripPropTest` to reproduce the bug and prevent a regression.

I reverted my change because I overcomplicated it last night. This fix is simpler than what I did last night

File: src/dbnode/encoding/proto/docs/encoding.md [hunk 24]
```diff
@@ -95,7 +95,7 @@ In practice, a Protbuf message can have any number of different fields; for perf
 Instead, multiple "logical" streams are interleaved on a per-write basis within one physical stream.
 The remainder of this section outlines the binary format used to accomplish this interleaving.
 
-The binary format begins with a stream header and then the remainder of the stream is a sequence of tuples in the form: `<per-write header, compressed timestamp, compressed custom encoded fields, Protobuf marshaled fields>`
+The binary format begins with a stream header and then the remainder of the stream is a sequence of tuples in the form: `<per-write header, compressed timestamp, compressed custom encoded fields, Protobuf marshalled fields>`
```
Comment (mway) [inline src/dbnode/encoding/proto/docs/encoding.md]: nit:

> [...] stream header, and the remainder [...]

File: src/dbnode/encoding/ostream.go [hunk 25]
```diff
@@ -155,10 +155,17 @@ func (os *ostream) WriteByte(v byte) {
 
 // WriteBytes writes a byte slice.
 func (os *ostream) WriteBytes(bytes []byte) {
+	if !os.hasUnusedBits() {
+		// If the stream is aligned on a byte boundary then all of the WriteByte()
+		// function calls and bit-twiddling can be skipped in favor of a single
+		// memcopy.
```
[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/encoding/ostream.go]: You're right, what I'm doing would bypass the pooling. I'll change it to call ensureCapacityFor() first.

[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/encoding/ostream.go]: I'm gonna stick with append because its harder to mess it up and as long as I call `ensureCapacityFor()` first it will be equivalent to a copy anyways.

Also thanks for pointing this out because I was forcing allocations by doing an append when I could use `ensureCapacityFor` to get a new byte slice from the pool if availabble

File: src/dbnode/encoding/ostream_test.go [hunk 26]
```diff
@@ -89,7 +89,7 @@ func testWriteBytes(t *testing.T, o OStream) {
 	b, _ := os.Rawbytes()
 	require.Equal(t, rawBytes, b)
 
-	require.Equal(t, 8, os.pos)
+	require.Equal(t, 0, os.pos)
```
[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/encoding/ostream_test.go]: its a quirk of the implementation of the `ostream` that both pos `0` and pos `8` mean that the last byte is completely full:

```golang
func (os *ostream) hasUnusedBits() bool {
	return os.pos > 0 && os.pos < 8
}
```

That said its better to not change stuff if there is no need and I was perhaps a bit hasty here by changing this test. I'll change my implementation to ensure that it ends up with `pos` set to 8 instead of 0

File: src/dbnode/encoding/proto/buffer_encode.go [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// This file is mostly copy-pasta from: https://github.com/jhump/protoreflect/blob/master/dynamic/codec.go
```
[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/encoding/proto/buffer_encode.go]: nothing, just copied the methods I needed

File: src/dbnode/encoding/proto/common.go [hunk 28]
```diff
@@ -141,11 +143,19 @@ var (
 	}
 )
 
+type marshalledField struct {
+	fieldNum   int32
+	marshalled []byte
+}
+
+type sortedMarshaledFields []marshalledField
```
[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/encoding/proto/common.go]: since I already refactored the entire package from `marshaled` to `marshalled` I'm gonna stick with `marshalled`. Will fix this

File: src/dbnode/encoding/proto/common.go [hunk 29]
```diff
@@ -167,57 +177,83 @@ type encoderBytesFieldDictState struct {
...
 	fields := schema.GetFields()
 	numCustomFields := numCustomFields(schema)
-	numProtoFields := len(fields) - numCustomFields
+	numNonCustomFields := len(fields) - numCustomFields
 
-	if cap(s) >= numCustomFields {
-		for i := range s {
-			s[i] = customFieldState{}
+	if cap(customFields) >= numCustomFields {
+		for i := range customFields {
```
[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/encoding/proto/common.go]: I specifically tried to avoid pointers since I can have a lot of these in memory at once so I'd rather not have the G.C chasing them.

In term of allocating, the for loop here basically boils down to a memclear since all of the (non pointer objects) will exist as one contiguous slice in memory and its just rolling through all of them and zeroing them out. Will not call into any `malloc` or G.C code

File: src/dbnode/encoding/proto/custom_marshal.go [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+func newCustomMarshaller() customFieldMarshaller {
+	return &customMarshaller{
+		buf: newCodedBuffer(nil),
+	}
+}
+
+func (m *customMarshaller) encFloat64(tag int32, x float64) {
+	if x == 0 {
```
[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/encoding/proto/custom_marshal.go]: took me a minute to decipher this haha. I assumed that this would actually force the compiler to make the 0 -> 0.0 but you maybe be right that this is a type assertion. Good catch

File: src/dbnode/encoding/proto/common.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (mway) [reply src/dbnode/encoding/proto/common.go]: afaict the GC is going to end up chasing most of these anyway because they're going to be promoted from stack to heap storage as the lifetime of non-garbage `customField` members outlives their creating scope by being stored in a heap-allocated container - so the question really becomes whether it's worth it to shave off the extra stack allocations (which are admittedly not very expensive, though it's more a question of frequency/volume here) for ostensibly no change in GC behavior.

am i missing something? do you think that the difference is negligible or, worse, actually detrimental?

[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/encoding/proto/common.go]: I guess what I'm saying is:

```golang
[]*customFieldState
```

in memory will be represented as a contiguous region in memory where each slot is 8 bytes (enough for a 64 bit pointer to the actual customFieldState which can be anywhere else on the heap.

```golang
[]customFieldState
```

will be a contiguous region in memory where each slot is equal to `unsafe.Sizeof(customFieldState{}`

If we have a slice of size 20, in both cases the G.C is going to have to do some work because the `customFieldState` itself has pointers that need to be chase, however, the latter doesn't have any pointers for the `customFieldState` which means thats 20 less heap allocated objects that the G.C needs to trace.

Basically by making it a slice of non-pointers you're forcing them all to be next to each other in memory and their are no long-lived pointers to any individual customFieldState.

As a result, when you loop through the slice and do `nonCustomFields[i] = customFieldState{}` you're not allocating anything new. You're just going to the slot in the slice and setting the space thats already been aside to a zero instance of the struct but nothing new had to be allocated from the heap, the space was already there, you just zero'd out all the fields.

[the remark this answers is not in the document]
Comment (mway) [reply src/dbnode/encoding/proto/common.go]: ah, yep, you're right. i was glossing over the contiguous/non-contiguous member tradeoff. thanks for explaining!

[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/encoding/proto/common.go]:
```golang
var dontOptimize []customFieldState

func BenchmarkResetCustomFieldState(b *testing.B) {
	dontOptimize = make([]customFieldState, 100)
	for i := 0; i < b.N; i++ {
		for i := range dontOptimize {
			dontOptimize[i] = customFieldState{}
		}
	}
	// Make sure it cant be optimized.
	fmt.Println(dontOptimize)
}
```

```
goos: darwin
goarch: amd64
pkg: github.com/m3db/m3/src/dbnode/encoding/proto
Running tool: /usr/local/bin/go test -benchmem -run=^$ github.com/m3db/m3/src/dbnode/encoding/proto -bench ^(BenchmarkResetCustomFieldState)$

BenchmarkResetCustomFieldState-8   	 2000000	       621 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/m3db/m3/src/dbnode/encoding/proto	1.953s
```
